### PR TITLE
add paypal payment support

### LINKED ISSUE #39: Add support for PayPal payments
we want to be able to accept payments and send info to the backend to change subscriptions of users

### PR DESCRIPTION
**Pull request by mattwr18**
Adds Premium membership subscription through PayPal
putting in this `PR` as it's already getting big, but there are still many other tasks that need to be done, which I will create issues for. Please create issues as you see fit @joaopapereira @Kachulio1 @FedericoEsparza 
#### Issue addressed

fixes #39

#### Screenshots (if appropriate):

screenshot at 2019-02-10 11 40 53

#### Testing

has been unit tested... there are some tests see `components/PayPalSuccess.test.js` where testing the function within `useEffect` proved difficult... also for the Subscriptions.test.js the second test fails if the first, not-so-good test, isn't present

### CODE REVIEW

File: src/components/App.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import React, { Component, Fragment } from 'react'
```
Comment (joaopapereira) [inline src/components/App.js:1]: This change should have came in a separated pr

File: src/containers/LogIn.js [hunk 2]
```diff
@@ -30,6 +33,7 @@ export class LogIn extends Component {
         })
       })
       .catch(() => {
+        console.log('error')
```
Comment (joaopapereira) [inline src/containers/LogIn.js:36]: We should remove this log from here.
Unless we want to save the error that happened in the console

File: src/containers/ProjectsList.js [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import React, { Component, Fragment } from 'react'
-import { Header, Card, Grid } from 'semantic-ui-react'
+import { Header, Card, Grid, Container } from 'semantic-ui-react'
```
Comment (joaopapereira) [inline src/containers/ProjectsList.js]: Is this change related to PayPal?

File: src/containers/UsersList.js [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import React, { Component, Fragment } from 'react'
-import { Header, Card, Grid } from 'semantic-ui-react'
+import { Header, Card, Grid, Container } from 'semantic-ui-react'
```
Comment (joaopapereira) [inline src/containers/UsersList.js]: Is this change related to PayPal?

File: src/helpers/createBillingAgreement.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import axios from 'axios'
```
Comment (joaopapereira) [inline src/helpers/createBillingAgreement.js:1]: Is this the correct path for this file? Sounds like it should live as part of the actions or something else since it is a http request. What is the rational behind placing it here?

File: src/helpers/executeBillingAgreement.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import axios from 'axios'
```
Comment (joaopapereira) [inline src/helpers/executeBillingAgreement.js:1]: Is this the correct path for this file? Sounds like it should live as part of the actions or something else since it is a http request. What is the rational behind placing it here?

File: src/tests/components/PayPalAgreementNew.test.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    createBillingAgreement: jest.fn()
+  }
+  beforeEach(() => {
+    wrapper = shallow(<PayPalAgreementNew {...props} />)
+  })
+
+  it('has a form to submit a new paypal agreement', () => {
+    wrapper.find('form').simulate('submit')
+    expect(props.createBillingAgreement).toHaveBeenCalled()
+  })
```
Comment (joaopapereira) [inline src/tests/components/PayPalAgreementNew.test.js:18]: What happens if an error occurs?

File: src/tests/components/PayPalSuccess.test.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    location: { search: '?plan=premium' }
+  }
+  beforeEach(() => {
+    wrapper = mount(<PayPalSuccess {...props} />)
+  })
+
+  it('renders a success segment', () => {
+    wrapper.setProps({ location: { search: '?plan=premiummob' } })
+    expect(wrapper.find('Segment')).toBeTruthy()
+  })
```
Comment (joaopapereira) [inline src/tests/components/PayPalSuccess.test.js:18]: What happens if an error occurs?

File: src/components/Subscriptions.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+import { connect } from 'react-redux'
+import { setLastLocation } from '../actions/setLastLocationAction'
+import { Header, Container, Segment, Grid } from 'semantic-ui-react'
+import PayPalAgreementNew from './PayPalAgreementNew'
+import createBillingAgreement from '../helpers/createBillingAgreement'
+import queryString from 'query-string'
+import '../assets/Subscriptions.css'
+
+export const Subscriptions = props => {
+  useEffect(() => {
```
Comment (joaopapereira) [inline src/components/Subscriptions.js:11]: I tried to access the URL for this component directly but I got a blank page

File: src/tests/components/Subscriptions.test.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import React from 'react'
+import { mount } from 'enzyme'
+import { Subscriptions } from '../../components/Subscriptions'
+
+describe('Subscriptions', () => {
```
Comment (joaopapereira) [inline src/tests/components/Subscriptions.test.js:5]: Not sure how useful this test is, I would use a Cypress test for this

**Review by joaopapereira — CHANGES_REQUESTED** [review]
Awesome job getting this PayPal stuff together. 
Please review my comments

File: src/components/App.js (hunk 1 above)
Comment (mattwr18) [reply src/components/App.js:1]: this change is in this `PR` as it is needed to get paypal working since we need to pass the cookies to the backend

File: src/components/Subscriptions.js (hunk 9 above)
Comment (mattwr18) [reply src/components/Subscriptions.js:11]: that is probably because you tried to access `/subscriptions/new`? you need to try to access `/subscriptions/new?plan=premium` as we would like this Subscriptions component to take care of premium, premiummob, and premiumf2f

File: src/containers/LogIn.js (hunk 2 above)
Comment (mattwr18) [reply src/containers/LogIn.js:36]: yep, can remove that

File: src/containers/ProjectsList.js (hunk 3 above)
Comment (mattwr18) [reply src/containers/ProjectsList.js]: no, it is related to error in the console from having everything wrapped in a Container in index.js

File: src/containers/UsersList.js (hunk 4 above)
Comment (mattwr18) [reply src/containers/UsersList.js]: same comment as above

File: src/helpers/createBillingAgreement.js (hunk 5 above)
Comment (mattwr18) [reply src/helpers/createBillingAgreement.js:1]: I don't know if there is a "correct" path... @Kachulio1 suggested to put it in the helpers directory, I think the actions directory is for actions that will dispatch info to the store in redux, as this doesn't do that and is just a helper function, it is in the helpers directory

File: src/helpers/executeBillingAgreement.js (hunk 6 above)
Comment (mattwr18) [reply src/helpers/executeBillingAgreement.js:1]: same comment as above

File: src/tests/components/PayPalAgreementNew.test.js (hunk 7 above)
Comment (mattwr18) [reply src/tests/components/PayPalAgreementNew.test.js:18]: we are not handling errors at the moment, hence why I said in PR comments that this still needs to have separate tickets to handle all the edge cases.

File: src/tests/components/PayPalSuccess.test.js (hunk 8 above)
Comment (mattwr18) [reply src/tests/components/PayPalSuccess.test.js:18]: same comment as above

File: src/tests/components/Subscriptions.test.js (hunk 10 above)
Comment (mattwr18) [reply src/tests/components/Subscriptions.test.js:5]: to quote myself above "there are some tests see `components/PayPalSuccess.test.js` where testing the function within `useEffect` proved difficult... also for the Subscriptions.test.js the second test fails if the first, not-so-good test, isn't present", although I believe in a subsequent `PR` I was able to get the second test to pass without the first

File: src/components/Subscriptions.js (hunk 9 above)
Comment (mattwr18) [reply src/components/Subscriptions.js:11]: Although, maybe like the agileventures site, we should have `subscriptions/new` default to premium sign up info

File: src/containers/ProjectsList.js (hunk 3 above)
Comment (mattwr18) [reply src/containers/ProjectsList.js]: if you merge in this `PR` first, it will remove it from the files changed https://github.com/AgileVentures/WebsiteOne-FE/pull/79

File: src/containers/UsersList.js (hunk 4 above)
Comment (mattwr18) [reply src/containers/UsersList.js]: https://github.com/AgileVentures/WebsiteOne-FE/pull/79

File: src/tests/components/PayPalAgreementNew.test.js (hunk 7 above)
Comment (mattwr18) [reply src/tests/components/PayPalAgreementNew.test.js:18]: have created this issue https://github.com/AgileVentures/WebsiteOne-FE/issues/80

File: src/tests/components/PayPalSuccess.test.js (hunk 8 above)
Comment (mattwr18) [reply src/tests/components/PayPalSuccess.test.js:18]: https://github.com/AgileVentures/WebsiteOne-FE/issues/80

File: src/helpers/createBillingAgreement.js (hunk 5 above)
Comment (joaopapereira) [reply src/helpers/createBillingAgreement.js:1]: Helpers do not feed the correct place to put it, but since I do not have a good alternative for now I think we will keep it

**Review by joaopapereira — APPROVED** [review]
Looks good, going to merge
